File: cmd/kingpin-web/web/js/police.js
// The law panel (#355): the police risk in a city, heat, the file,
// pressure and the cash exposure together, each said in a clause, the
// way the TUI's POLICE section says it. A pure function of the view, so
// the page draws it and the Node test runs it; it touches no DOM.

const pct = (x) => `${Math.round(x * 100)}%`;
const money = (n) => (n < 0 ? "-$" : "$") + Math.abs(Math.round(n)).toLocaleString("en-US");
const plural = (n, word) => `${n} ${word}${n === 1 ? "" : "s"}`;

// rungWord is what a rung does, in one clause.
export function rungWord(r) {
  switch (r.level) {
    case "patrol":
      return `sales capped near ${pct(r.cap || 0)} of demand for ${plural(r.cap_days || 0, "day")}`;
    case "arrest":
      return "you are arrested: the run ends";
    case "taskforce":
      return `the feds seize an asset and take ${pct(r.stock_loss || 0)} of the stock and ${pct(r.cash_loss || 0)} of the dirty cash`;
    default:
      return `takes ${pct(r.stock_loss || 0)} of the stock and ${pct(r.cash_loss || 0)} of the dirty cash`;
  }
}

const rungName = (level) => (level === "taskforce" ? "task force" : level);

// fileWord is the header's `3/6`, or the page count alone with no file.
export function fileWord(v) {
  const line = v.law && v.law.arrest_line;
  return line > 0 ? `${v.you.evidence}/${line}` : String(v.you.evidence);
}

// policeLines is the panel for a city: a list of {text, warn} lines,
// warn set on what is close.
export function policeLines(v, cityId) {
  const c = v.cities.find((x) => x.id === cityId);
  if (!c) return [];
  const out = [];
  const say = (text, warn = false) => out.push({ text, warn });
  const ladder = c.ladder || [];
  const next = ladder.find((r) => c.heat < r.line);
  say(`Heat ${Math.round(c.heat)}/100${next ? `: ${Math.ceil(next.line - c.heat)} to the ${rungName(next.level)}` : ""}.`, !!next && next.line - c.heat <= 10);
  for (const r of ladder) {
    const pages = r.pages ? `, ${plural(r.pages, "page")} if you sold that day` : "";
    say(`${rungName(r.level)} ${Math.round(r.line)}: ${rungWord(r)}${pages}.`, c.heat >= r.line);
  }
  const line = v.law.arrest_line;
  if (line > 0) {
    const left = Math.max(0, line - v.you.evidence);
    say(`The file: ${v.you.evidence}/${line} pages, ${plural(left, "page")} from an indictment. A bust on a day you sold files pages; a quiet day files nothing.`, left <= 2);
  }
  say(`Pressure ${Math.round(c.pressure)}, goodwill ${Math.round(c.goodwill)}: pressure lowers the patrol, raid and arrest lines and tightens a patrol; goodwill wears it down.`);
  const exp = v.law.exposure_line;
  if (exp > 0) {
    const cover = v.law.cover > 0 ? ` (${money(exp - v.law.cover)} plus ${money(v.law.cover)} your fronts cover)` : "";
    const past = v.you.dirty_cash > exp;
    say(`Dirty cash ${money(v.you.dirty_cash)} of a ${money(exp)} line${cover}: ${past ? "over it, the pile draws heat every day" : "under it, the pile draws nothing"}.`, past);
  }
  if (c.response) {
    say(`Estimate: a cop says the ${rungName(c.response)} from day ${c.response_day}, ${pct(c.response_sure || 0)} sure.`);
  } else {
    say("No word from inside: nothing tells you when they move.");
  }
  return out;
}

File: internal/ui/police.go
// The POLICE section (#355): the police risk in a city in one place,
// heat, the file, pressure and the cash exposure, each said in a clause.
// The dashboard's pane leads with it while the arrows are past the
// product table on HEAT, CASH and LAW (onPolice), and the map's carries
// it for the city shown, under the corner.

package ui

import (
	"fmt"
	"strings"

	"github.com/theclifmeister/kingpin/internal/content"
	"github.com/theclifmeister/kingpin/internal/format"
	"github.com/theclifmeister/kingpin/internal/game"
	"github.com/theclifmeister/kingpin/internal/ui/theme"
)

// dashboardMove is the dashboard's arrows: down the product table and,
// past its last row, onto the police (HEAT, CASH and LAW), whose
// section then leads the pane; up from there is the table again.
func (m *Model) dashboardMove(dy int) {
	switch {
	case m.onPolice && dy < 0:
		m.onPolice = false
	case !m.onPolice && dy > 0 && m.cursor >= len(m.w.Products)-1:
		m.onPolice = true
	case !m.onPolice:
		m.productMove(dy)
	}
}

// policeTitle is a dashboard panel's title, marked while the arrows are
// on the police.
func (m *Model) policeTitle(title string) string {
	if m.onPolice && m.screen == screenDashboard {
		return "▸ " + title
	}
	return title
}

// rungName is a rung as the pane's label column holds it: the task
// force is the feds.
func rungName(level string) string {
	if level == content.TaskForce {
		return "feds"
	}
	return level
}

// rungWords is what a rung does, as a row's value: the share of the
// stock and the dirty cash it takes, a patrol's cap and its days, the
// feds' asset, the arrest's end. Two lines for the feds.
func rungWords(r content.ResponseConfig) []string {
	take := fmt.Sprintf("%s stock, %s cash", format.Pct(r.StockLoss, 0), format.Pct(r.CashLoss, 0))
	switch r.Level {
	case content.Patrol:
		return []string{fmt.Sprintf("sales ~%s for %dd", format.Pct(r.Cap, 0), r.CapDays)}
	case content.Arrest:
		return []string{"a warrant, then a cell"} // #475: served the next night on a sale or the heat held
	case content.TaskForce:
		return []string{"an asset, and", take}
	}
	return []string{take}
}

// policeSection is the police risk in a city: the heat and how far the
// next rung is, each rung's line and what it takes, the file and what
// fills it, the pressure and the goodwill and what they move, the dirty
// pile against the line with the fronts' cover, and the cop's word as an
// estimate. The lines are heat.Sim's (Rungs, EvidenceArrest,
// ExposureLine, Cover), so the pane and the dice agree; the forecast is
// the file's (game.Known), never the truth.
func (m *Model) policeSection(city *game.City) section {
	w := m.w
	rungs := m.rules.Heat.Rungs(w, city)
	var lines []string

	// The heat, and the first rung it is under.
	heat := fmt.Sprintf("%.0f", city.Heat)
	for _, r := range rungs {
		if city.Heat < r.Threshold {
			heat += fmt.Sprintf(" · %s in %.0f", rungName(r.Level), r.Threshold-city.Heat)
			break
		}
	}
	lines = append(lines, row("heat", heatStyle(city.Heat).Render(heat)))
	next := true
	for _, r := range rungs {
		style := theme.Subtle
		switch {
		case city.Heat >= r.Threshold:
			style = theme.Bad
		case next:
			style, next = theme.Warning, false
		}
		label := fmt.Sprintf("%s %.0f", rungName(r.Level), r.Threshold)
		for _, v := range rungWords(r) {
			lines = append(lines, row(label, style.Render(v)))
			label = ""
		}
	}

	// The file: how thick, how far from the line, and what adds a page.
	if arrest := m.rules.Heat.EvidenceArrest(w); arrest > 0 {
		left := max(0, arrest-w.Heat.Evidence)
		style := theme.Subtle
		if left <= 2 {
			style = theme.Bad
		}
		lines = append(lines, row("file", style.Render(fmt.Sprintf("%d/%d · %s to go", w.Heat.Evidence, arrest, plural(left, "page")))))
		var pages []string
		for _, r := range rungs {
			if r.Evidence > 0 {
				name := strings.ToUpper(r.Level[:1]) + r.Level[1:]
				if len(pages) > 0 {
					name = r.Level
				}
				if r.Level == content.TaskForce {
					name = "feds"
				}
				pages = append(pages, fmt.Sprintf("%s %d", name, r.Evidence))
			}
		}
		note := "A quiet day files nothing."
		if len(pages) > 0 {
			note = strings.Join(pages, ", ") + " pages, if you sold; a quiet day files nothing."
		}
		lines = append(lines, m.wrapped(theme.Subtle, note)...)
	} else {
		lines = append(lines, row("file", theme.Subtle.Render("no case can be made")))
	}

	// Pressure and goodwill: what the one moves and the other takes off.
	pg := fmt.Sprintf("%.0f", city.Pressure)
	if city.Goodwill > 0 {
		pg += theme.Good.Render(fmt.Sprintf(" · goodwill %.0f", city.Goodwill))
	}
	lines = append(lines, row("pressure", theme.Bad.Render(pg)))
	fx := m.cfg.Law.Effects
	lt := m.rules.Law.Tuning()
	note := fmt.Sprintf("Lowers the police lines %s and a patrol's cap %s; fades to %.0f.",
		format.Pct(1-content.Cut(city.Pressure, fx.PressureThresholdCut), 0), format.Pct(1-content.Cut(city.Pressure, fx.PressureCapCut), 0), lt.Baseline)
	if city.Goodwill > 0 {
		note += fmt.Sprintf(" Goodwill takes %.1f a day.", lt.GoodwillCut*city.Goodwill/100)
	}
	lines = append(lines, m.wrapped(theme.Subtle, note)...)

	// The dirty pile against the line: the threshold and the cover.
	if line := m.rules.Heat.ExposureLine(w); line > 0 {
		dirty := w.Player.DirtyCash
		style := theme.Gold
		if dirty > line {
			style = theme.Warning
		}
		lines = append(lines, row("dirty", style.Render(fmt.Sprintf("%s of %s", cash(dirty), cash(line)))))
		if cover := m.rules.Heat.Cover(w); cover > 0 {
			lines = append(lines, row("", theme.Subtle.Render(fmt.Sprintf("%s + %s fronts", cash(line-cover), cash(cover)))))
		}
		if dirty > line {
			lines = append(lines, row("", theme.Warning.Render("over: heat every day")))
		} else {
			lines = append(lines, row("", theme.Subtle.Render("under: draws nothing")))
		}
	} else {
		lines = append(lines, row("dirty", theme.Subtle.Render("pile draws nothing")))
	}

	// The forecast: the cop's word, an estimate, or none.
	k := m.known()
	if f, ok := k.Fact(city.ID, game.FactResponse); ok {
		level, day, _ := k.Response(city.ID)
		lines = append(lines,
			row("estimate", theme.Warning.Render(fmt.Sprintf("%s from day %d", rungName(level), day))),
			row("", theme.Subtle.Render(fmt.Sprintf("a cop · %s sure", format.Pct(f.Now(w.Day), 0)))))
	} else {
		lines = append(lines, row("estimate", theme.Subtle.Render("no word from inside")))
	}
	return section{"POLICE · " + strings.ToUpper(city.Name), lines}
}
